### PR TITLE
Redire to the last URI after login:

### PR DESCRIPTION
**Pull request by razdob15**
We used the OIDC `state` query parameter to return to the last page in NiFi after login.

### CODE REVIEW

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/OIDCAccessResource.java [hunk 1]
```diff
@@ -175,9 +175,13 @@ public void oidcCallback(@Context HttpServletRequest httpServletRequest, @Contex
             }
 
             // redirect to the name page
-            httpServletResponse.sendRedirect(getNiFiUri());
-        } else {
-            // remove the oidc request cookie
```
Comment (razdob15) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/OIDCAccessResource.java:180]: Keep this comment before the relevant line

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/OIDCAccessResource.java [hunk 2]
```diff
@@ -175,9 +175,13 @@ public void oidcCallback(@Context HttpServletRequest httpServletRequest, @Contex
...
+            if (successfulOidcResponse.getState().getValue().contains(".")) {
```
Comment (razdob15) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/OIDCAccessResource.java]: Add comments about the `if` block.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/OIDCAccessResource.java [hunk 3]
```diff
@@ -259,7 +263,7 @@ public void oidcLogout(@Context HttpServletRequest httpServletRequest, @Context
             case REVOKE_ACCESS_TOKEN_LOGOUT:
             case ID_TOKEN_LOGOUT:
                 // Make a request to the IdP
-                URI authorizationURI = oidcRequestAuthorizationCode(httpServletResponse, getOidcLogoutCallback());
+                URI authorizationURI = oidcRequestAuthorizationCode(httpServletResponse, getOidcLogoutCallback(), "");
```
Comment (razdob15) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/OIDCAccessResource.java:270]: Add a comment explains the last new variable.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/OIDCAccessResource.java [hunk 4]
```diff
@@ -402,15 +406,20 @@ public void oidcLogoutCallback(@Context HttpServletRequest httpServletRequest, @
      * @param callback the OIDC callback URI
      * @return the authorization URI
      */
-    private URI oidcRequestAuthorizationCode(@Context HttpServletResponse httpServletResponse, String callback) {
+    private URI oidcRequestAuthorizationCode(@Context HttpServletResponse httpServletResponse, String callback, String uri) {
```
Comment (razdob15) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/OIDCAccessResource.java]: - Rename `uri` variable. 
- Add documentation about the new variable.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/OIDCAccessResource.java [hunk 5]
```diff
@@ -402,15 +406,20 @@ public void oidcLogoutCallback(@Context HttpServletRequest httpServletRequest, @
...
         final String oidcRequestIdentifier = UUID.randomUUID().toString();
         applicationCookieService.addCookie(getCookieResourceUri(), httpServletResponse, ApplicationCookieName.OIDC_REQUEST_IDENTIFIER, oidcRequestIdentifier);
         final State state = oidcService.createState(oidcRequestIdentifier);
+
+        if (uri.startsWith("/")) {
```
Comment (razdob15) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/OIDCAccessResource.java]: - [x] Explain why in a comment. 
- [x] Check for a better function to remove the first character if it's a `/`.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-security/src/main/java/org/apache/nifi/web/security/oidc/OidcService.java [hunk 6]
```diff
@@ -177,7 +177,7 @@ public boolean isStateValid(final String oidcRequestIdentifier, final State prop
                 stateLookupForPendingRequests.invalidate(oidcRequestIdentifierKey);
             }
 
-            return state != null && IdentityProviderUtils.timeConstantEqualityCheck(state.getValue(), proposedState.getValue());
+            return state != null && IdentityProviderUtils.timeConstantEqualityCheck(state.getValue(), proposedState.getValue().split("\\.")[0]);
```
Comment (razdob15) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-security/src/main/java/org/apache/nifi/web/security/oidc/OidcService.java]: Explain in a comment that we take the random part from the state, which is before the first dot.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-canvas.js [hunk 7]
```diff
@@ -944,7 +944,7 @@
                         // there is no anonymous access and we don't know this user - open the login page which handles login/registration/etc
                         if (xhr.status === 401) {
                             nfAuthorizationStorage.removeToken();
-                            window.location = '../nifi/login';
+                            window.location = '../nifi/login?' + window.location.href.split('/nifi')[1];
```
Comment (razdob15) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-canvas.js:949]: Explain this line in a comment.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/OIDCAccessResource.java [hunk 8]
```diff
@@ -175,9 +175,13 @@ public void oidcCallback(@Context HttpServletRequest httpServletRequest, @Contex
             }
 
             // redirect to the name page
-            httpServletResponse.sendRedirect(getNiFiUri());
-        } else {
-            // remove the oidc request cookie
+            if (successfulOidcResponse.getState().getValue().contains(".")) {
+                String redirectPagePathBase64 = successfulOidcResponse.getState().getValue().substring(successfulOidcResponse.getState().getValue().lastIndexOf(".") + 1);
+                httpServletResponse.sendRedirect(getNiFiUri() + new String(Base64.getDecoder().decode(redirectPagePathBase64)));
+            } else {
```
Comment (Cyben) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/OIDCAccessResource.java]: Change to clean code standard

**Comment by Cyben** [comment]
Hey @razdob15 ,
We should consider simulating some AuthZ and security checks.
For examaple,

What happens if someone is trying to connect from the beginning to a forbidden route?

What happens if someone is trying to insert some valuable data after the "." in the state parameters or a long string? Maybe some overflow or so?

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/OIDCAccessResource.java (hunk 8 above)
Comment (razdob15) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/OIDCAccessResource.java]: What do you mean?

Comment (Cyben) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/OIDCAccessResource.java]: Build the url inside the if and else statements, and move the redirect call httpServletResponse.sendRedirect under the if else condition (line 184 for example)

**Pull request merged**